subgraphs/across-v2/src/handlers/liquidity.ts

```diff
@@ -0,0 +1,167 @@
+import { Address, BigInt, Bytes, log } from "@graphprotocol/graph-ts";
+import {
+  HubPool,
+  LiquidityAdded,
+  LiquidityRemoved,
+} from "../../generated/HubPool/HubPool";
+import { SDK } from "../sdk/protocols/bridge";
+import { BridgePoolType } from "../sdk/protocols/bridge/enums";
+import {
+  ACROSS_HUB_POOL_CONTRACT,
+  MAINNET_BRIDGE_CONFIG,
+  Pricer,
+  TokenInit,
+  getTokenBalance,
+} from "../util";
+import { _OutputTokenToPool } from "../../generated/schema";
+import { _ERC20 } from "../../generated/SpokePool1/_ERC20";
+
+export function handleLiquidityAdded(event: LiquidityAdded): void {
+  const sdk = SDK.initializeFromEvent(
+    MAINNET_BRIDGE_CONFIG,
+    new Pricer(event.block),
+    new TokenInit(),
+    event
+  );
+
+  // input token
+  const token = sdk.Tokens.getOrCreateToken(event.params.l1Token);
+
+  // output token
+  const hubPoolContract = HubPool.bind(
+    Address.fromString(ACROSS_HUB_POOL_CONTRACT)
+  );
+  const hubPoolContractCall = hubPoolContract.try_pooledTokens(
+    event.params.l1Token
+  );
+  let outputTokenAddress: Address;
+  if (hubPoolContractCall.reverted) {
+    log.info("[HubPool:pooledTokens()] get LP Token call reverted", []);
+    outputTokenAddress = event.params.l1Token;
+  } else {
+    outputTokenAddress = hubPoolContractCall.value.getLpToken();
+  }
+  const outputToken = sdk.Tokens.getOrCreateToken(outputTokenAddress);
+
+  // pool
+  const poolId = event.params.l1Token.concat(Bytes.fromUTF8("liquidity"));
+  const pool = sdk.Pools.loadPool<string>(poolId);
+
+  if (!pool.isInitialized) {
+    pool.initialize(token.name, token.symbol, BridgePoolType.LIQUIDITY, token);
+    pool.pool.outputToken = outputToken.id;
+    pool.save();
+  }
+
+  // output token supply
+  let outputTokenSupply: BigInt;
+  const erc20 = _ERC20.bind(outputTokenAddress);
+  const outputTokenSupplyResult = erc20.try_totalSupply();
+  if (outputTokenSupplyResult.reverted) {
+    log.info(
+      "[_ERC20:tokenSupply()] retrieve outputTokenSupply for LP pool call reverted",
+      []
+    );
+  } else {
+    outputTokenSupply = outputTokenSupplyResult.value;
+    pool.setOutputTokenSupply(outputTokenSupply);
+  }
+
+  // output token price
+  pool.pool.outputTokenPriceUSD = token.lastPriceUSD;
```

Comment:
Shouldn't output token price be TVL / output token supply?

Reply:
Prices for output (LP) tokens aren't available. Since across's LP token minted/burned is 1:1 with the input token deposited/withdrawn, we can use the input token price as a substitute for output token price.

Reply:
Prices for output (LP) tokens aren't available and there were no contract read methods for fetching the exchangeRate. Since across's LP token minted/burned is almost 1:1 with the input token deposited/withdrawn, I went ahead with the input token price as a substitute for output token price.

Reply:
Thinking further, we could derive the exchangeRate (`exchangeRate = l1Tokens / lpTokensMinted`) and then multiple input token price to derive the output token price.

Reply:
I had ignored exchangeRateCurrent() method earlier because this was a write contractual method. I didn't realize we could call write methods (this one is non-payable type) in a subgraph. So, we do have two options now: a) `exchangeRate = l1Tokens / lpTokensMinted` b) call `exchangeRateCurrent()`. 

Verified they return same amounts upto 15 decimals. I'll go ahead with (a) to avoid contract calls.

Reply:
What do you think [about this change](https://github.com/messari/subgraphs/pull/1777/commits/aa46df830473bec3653effadad063f5c05e6e80e)?

Reply:
This might also work. I was imagining 

```typescript
market.outputTokenPriceUSD = market.inputTokenBalanceUSD / market.outputTokenBalance
```

That should be the same thing, and no need to make a contract call for the exchange rate.

Reply:
I see. We should be good to go with the current implementation. Doesn't make contract call and avoids relying on data calculated by sdk (observed null wasm issues with the input token price).